include/gridtools/boundary_conditions/apply_gpu.hpp

```diff
@@ -185,96 +181,78 @@ namespace gridtools {
-        };
-    } // namespace _impl
-
-/** The following macro substitute the code to apply the boundary
-    function to the boundary portion identified by a particular
-    direction. The thread ids should be permuted since the shape
-    of the boundary portion may be transposed w.r.t. the kernel
-    configuration.
- */
-#define GT_RUN_BC_ON(x, y, z)                                                                                      \
-    if (predicate(direction<x, y, z>())) {                                                                         \
-        auto const &shape = conf.shape(static_cast<int>(x) + 1, static_cast<int>(y) + 1, static_cast<int>(z) + 1); \
-        if ((th[0] < shape.max()) && (th[1] < shape.median()) && (th[2] < shape.min())) {                          \
-            boundary_function(direction<x, y, z>{},                                                                \
-                data_views...,                                                                                     \
-                th[shape.perm(0)] + shape.start(0),                                                                \
-                th[shape.perm(1)] + shape.start(1),                                                                \
-                th[shape.perm(2)] + shape.start(2));                                                               \
-        }                                                                                                          \
-    }                                                                                                              \
-    static_assert(true, " ")
-
-    /**
-       @brief kernel to appy boundary conditions to the data fields requested
-     */
-    template <typename BoundaryFunction, typename Halos, typename... DataViews>
-    __global__ void loop_kernel(BoundaryFunction const boundary_function,
-        _impl::precomputed_pred const predicate,
-        _impl::kernel_configuration const conf,
-        Halos const halos,
-        DataViews const... data_views) {
-        array<uint_t, 3> th{blockIdx.x * blockDim.x + threadIdx.x,
-            blockIdx.y * blockDim.y + threadIdx.y,
-            blockIdx.z * blockDim.z + threadIdx.z};
-
-        GT_RUN_BC_ON(minus_, minus_, minus_);
-        GT_RUN_BC_ON(minus_, minus_, zero_);
-        GT_RUN_BC_ON(minus_, minus_, plus_);
-
-        GT_RUN_BC_ON(minus_, zero_, minus_);
-        GT_RUN_BC_ON(minus_, zero_, zero_);
-        GT_RUN_BC_ON(minus_, zero_, plus_);
-
-        GT_RUN_BC_ON(minus_, plus_, minus_);
-        GT_RUN_BC_ON(minus_, plus_, zero_);
-        GT_RUN_BC_ON(minus_, plus_, plus_);
-
-        GT_RUN_BC_ON(zero_, minus_, minus_);
-        GT_RUN_BC_ON(zero_, minus_, zero_);
-        GT_RUN_BC_ON(zero_, minus_, plus_);
-
-        GT_RUN_BC_ON(zero_, zero_, minus_);
-
-        GT_RUN_BC_ON(zero_, zero_, plus_);
+            dim3 blocks() const {
```

Comment:
maybe rename to `required_blocks` or `compute_required_blocks`

Reply:
Renamed blocks() and threads_per_block(), hope it’s more clear now.